File: 240704/k개의 벽 없애기/remove-k-walls.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>
#include <tuple>
#include <climits>
using namespace std;

#define MAX_N 100
#define DIR_NUM 4

int n, k;
int step[MAX_N][MAX_N];
bool visited[MAX_N][MAX_N], grid[MAX_N][MAX_N];
int r1, c1, r2, c2;

vector<pair<int, int>> wall_pos, selected_wall;
queue<pair<int, int>> q;

int ans = INT_MAX;

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

bool CanGo(int x, int y) {
    return InRange(x, y) && !visited[x][y] && !grid[x][y];
}

void Push(int x, int y, int num) {
    step[x][y] = num;
    visited[x][y] = 1;
    q.push(make_pair(x, y));
}

void BFS() {
    int dx[DIR_NUM] = {-1, 1, 0, 0};
    int dy[DIR_NUM] = {0, 0, -1, 1};

    while(!q.empty()) {
        int x, y;
        tie(x, y) = q.front();
        q.pop();

        for(int d=0; d<DIR_NUM; d++) {
            int nx = x+dx[d], ny = y+dy[d];

            if(CanGo(nx, ny)) {
                Push(nx, ny, step[x][y]+1);
            }
        }
    }
}

int Calc() {
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            step[i][j] = 0;
            visited[i][j] = 0;
        }
    }

    for(int i=0; i<selected_wall.size(); i++) {
        int x, y;
        tie(x, y) = selected_wall[i];
        grid[x][y] = 0;
    }

    Push(r1, c1, 0);
    BFS();

    for(int i=0; i<selected_wall.size(); i++) {
        int x, y;
        tie(x, y) = selected_wall[i];
        grid[x][y] = 1;
    }

    if(visited[r2][c2])
        return step[r2][c2];
    else
        return INT_MAX;
}

void FindMinMove(int idx, int cnt) {
    if (idx == wall_pos.size()) {
        if(cnt == k) {
            ans = min(ans, Calc());
        }
        return;
    }

    selected_wall.push_back(wall_pos[idx]);
    FindMinMove(idx+1, cnt+1);
    selected_wall.pop_back();

    FindMinMove(idx+1, cnt);
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>k;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
            if(grid[i][j])
                wall_pos.push_back(make_pair(i, j));
        }
    }

    cin >> r1>>c1>>r2>>c2;
    r1--; r2--; c1--; c2--;

    FindMinMove(0, 0);

    if(ans == INT_MAX)
        cout << -1;
    else 
        cout << ans;

    return 0;
}

File: 250704/K개의 벽 없애기/remove-k-walls.py
import sys
from collections import deque

INT_MAX = sys.maxsize

n, k = tuple(map(int, input().split()))
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

wall_pos = []
selc_pos = []

r1, c1 = tuple(map(int, input().split()))
r2, c2 = tuple(map(int, input().split()))

r1, c1, r2, c2 = r1-1, c1-1, r2-1, c2-1

q = deque()
visited = [[-1]*n for _ in range(n)]
ans = INT_MAX

def can_go(x, y):
    return 0<=x<n and 0<=y<n and visited[x][y] == -1 and not grid[x][y]


def bfs():
    global visited

    dxs, dys = [-1, 0, 1, 0], [0, 1, 0, -1]

    visited = [[-1]*n for _ in range(n)]
    q.append((r1, c1))
    visited[r1][c1] = 0

    for x, y in selc_pos:
        grid[x][y] = 0

    while q:
        x, y = q.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x+dx, y+dy
            
            if can_go(nx, ny):
                q.append((nx, ny))
                visited[nx][ny] = visited[x][y] + 1

    
    for x, y in selc_pos:
        grid[x][y] = 1


def find_min_dist(idx, cnt):
    global ans

    if cnt == k:
        bfs()
        if visited[r2][c2] != -1:
            ans = min(ans, visited[r2][c2])
        return
    
    if idx >= len(wall_pos):
        return
    
    selc_pos.append(wall_pos[idx])
    find_min_dist(idx+1, cnt+1)
    selc_pos.pop()

    find_min_dist(idx+1, cnt)

for i in range(n):
    for j in range(n):
        if grid[i][j]:
            wall_pos.append((i, j))


find_min_dist(0, 0)

print(ans if ans != INT_MAX else -1)

